### PR TITLE
Fix(rome_js_formatter): named import

### LINKED ISSUE #2441: Named imports formatting
Prettier breaks the named imports if the whole `import` statements exceed the line width, Rome only does so if the named imports exceed the line width [Playground](https://play.rome.tools/?lineWidth=80&indentStyle=space&quoteStyle=double&indentWidth=2&typescript=true&jsx=true&sourceType=module#aQBtAHAAbwByAHQAIABzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwAgAGYAcgBvAG0AIAAnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnACcACgBpAG0AcABvAHIAdAAgAHsAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAMQB9ACAAZgByAG8AbQAgACcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAJwAKAGkAbQBwAG8AcgB0ACAAYQAyADMANAA1ADUANgA2ADYANgA2ADYANgAsACAAewBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwAyAH0AIABmAHIAbwBtACAAJwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwAnAAoAaQBtAHAAbwByAHQAIAB7AGEAMgAzADQANQA1ADYANgA2ADYANgA2ADYAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQALAAgAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnADMAfQAgAGYAcgBvAG0AIAAnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnACcACgA=)

## Input

```ts
import somethingSuperLongsomethingSuperLong from 'somethingSuperLongsomethingSuperLongsomethingSuperLong'
import {somethingSuperLongsomethingSuperLong1} from 'somethingSuperLongsomethingSuperLongsomethingSuperLong'
import a234556666666, {somethingSuperLongsomethingSuperLong2} from 'somethingSuperLongsomethingSuperLongsomethingSuperLong'
import {a234556666666dddddddddddddddddd, somethingSuperLongsomethingSuperLong3} from 'somethingSuperLongsomethingSuperLongsomethingSuperLong'

```

## Prettier

```ts
import somethingSuperLongsomethingSuperLong from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
import { somethingSuperLongsomethingSuperLong1 } from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
import a234556666666, {
  somethingSuperLongsomethingSuperLong2,
} from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
import {
  a234556666666dddddddddddddddddd,
  somethingSuperLongsomethingSuperLong3,
} from "somethingSuperLongsomethingSuperLongsomethingSuperLong";

```

## Rome

```ts
import somethingSuperLongsomethingSuperLong from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
import { somethingSuperLongsomethingSuperLong1 } from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
import a234556666666, { somethingSuperLongsomethingSuperLong2 } from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
import {
  a234556666666dddddddddddddddddd,
  somethingSuperLongsomethingSuperLong3,
} from "somethingSuperLongsomethingSuperLongsomethingSuperLong";

```

## Expected

Rome's formatting to match Prettier's

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
## Summary
1. Fixed #2441 ,   

main: 
```bash
**File Based Average Prettier Similarity**: 71.96%  
**Line Based Average Prettier Similarity**: 65.15%  
```
pr: 
```bash
**File Based Average Prettier Similarity**: 71.97%  
**Line Based Average Prettier Similarity**: 65.16%  
```

## Test Plan
see the snapshot testing result, which should match the prettier formatting result of the linked issue

### CODE REVIEW

**Comment by ematipico** [comment]
@IWANABETHATGUY you could please add a comment to the issue you want to try to fix? Doing so, other people know that you're working on it and there won't be double PRs.

**Comment by IWANABETHATGUY** [comment]
> 

Yeah, I forget about that, thanks for reminding me.

File: crates/rome_formatter/src/lib.rs [hunk 1]
```diff
@@ -179,7 +179,18 @@ impl FormatOptions {
             ..Self::default()
         }
     }
-
+    pub fn with_line_width(mut self, line_width: LineWidth) -> Self {
```
Comment (IWANABETHATGUY) [inline crates/rome_formatter/src/lib.rs:182]: add builder style api , could be convenient when debugging the format code, you could write some code like this
```rust
 println!(
        "{}",
        Formatted::new(elements, FormatOptions::default().with_line_width(113.try_into().unwrap()))
            .print()
            .as_code()
    );
```

**Comment by IWANABETHATGUY** [comment]
!bench_formatter

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 2]
```diff
@@ -26,19 +30,55 @@ impl FormatNode for JsImportNamedClause {
...
...lse&jsx=false&sourceType=module#aQBtAHAAbwByAHQAIABzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwAgAGYAcgBvAG0AIAAnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnAHMAbwBtAGUAdABoAGkAbgBnAFMAdQBwAGUAcgBMAG8AbgBnACcACgBpAG0AcABvAHIAdAAgAHsAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwB9ACAAZgByAG8AbQAgACcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAJwAKAGkAbQBwAG8AcgB0ACAAYQAsACAAewBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwAyAH0AIABmAHIAbwBtACAAJwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwBzAG8AbQBlAHQAaABpAG4AZwBTAHUAcABlAHIATABvAG4AZwAnAAoAaQBtAHAAbwByAHQAIAB7AGEAMgAsACAAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAMwB9ACAAZgByAG8AbQAgACcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAcwBvAG0AZQB0AGgAaQBuAGcAUwB1AHAAZQByAEwAbwBuAGcAJwAKAA==
```
Comment (cpojer) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:44]: Let's not embed long urls like this, there is a high chance that this will get outdated.

Instead, maybe link to this PR or a GitHub issue that contains the link to the playground? That one is mutable, so we could edit the link in the future.

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 3]
```diff
@@ -26,19 +30,54 @@ impl FormatNode for JsImportNamedClause {
...
+                    group_elements(name),
+                    space_token(),
+                    from,
+                    space_token(),
+                    source,
+                    assertion,
+                ] // )
+            } // } else {
+              //     format_elements![name, space_token(), from, space_token(), source, assertion,]
+              // }
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:80]: Delete commented out code?

File: crates/rome_js_formatter/src/js/module/named_import_specifiers.rs [hunk 4]
```diff
@@ -14,6 +17,24 @@ impl FormatNode for JsNamedImportSpecifiers {
...
+                    },
+                    leading_trivia
+                ])
+            },
+            &r_curly_token?,
+        )
+        // formatter.format_delimited_soft_block_indent(
+        //     &l_curly_token?,
+        //     &r_curly_token?,
+        // )
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/module/named_import_specifiers.rs:38]: Delete commented out code?

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 5]
```diff
@@ -26,19 +30,54 @@ impl FormatNode for JsImportNamedClause {
...
+            let named_import = named_import?;
+            let syntax_node = named_import.syntax();
+            match syntax_node.kind() {
+                JsSyntaxKind::JS_NAMED_IMPORT_SPECIFIERS => {
+                    let specifiers =
+                        JsNamedImportSpecifiers::unwrap_cast(syntax_node.clone()).as_fields();
+                    specifiers.specifiers.syntax().children().count() > 1 || !default.is_empty()
+                }
+                _ => true,
+            }
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:54]: You can directly work on the `AstNode`s here, no need to use the `SyntaxNode`

```suggestion
            match named_improt? {
                JsAnyImportName::JsNamedImportSpecifiers(named_specifiers) => {
                    named_specifiers.specifiers.len() > 1 || !default.is_empty()
                }
                _ => true,
            }
```

Also, the `AstSeparatedList` offers a `len` method.

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 6]
```diff
@@ -26,19 +30,54 @@ impl FormatNode for JsImportNamedClause {
             format_elements![specifier, space_token()]
         })?;
         let from = from_token.format(formatter)?;
+
         let name = named_import.format(formatter)?;
         let assertion = assertion.format_with_or_empty(formatter, |assertion| {
             format_elements![space_token(), assertion]
         })?;
+        // only wrap the name_and_rest elements when named_specifiers is bigger than 1,
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:38]: The comment doesn't mention the case where the `default ` is empty.

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 7]
```diff
@@ -26,19 +30,54 @@ impl FormatNode for JsImportNamedClause {
...
+        // ```
+        // although this import has been exceeding the max line length, prettier still keep them all in one line
+        let need_to_wrap_group_elements = {
+            let named_import = named_import?;
+            let syntax_node = named_import.syntax();
+            match syntax_node.kind() {
+                JsSyntaxKind::JS_NAMED_IMPORT_SPECIFIERS => {
+                    let specifiers =
+                        JsNamedImportSpecifiers::unwrap_cast(syntax_node.clone()).as_fields();
+                    specifiers.specifiers.syntax().children().count() > 1 || !default.is_empty()
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:51]: We should avoid inspecting the formatter IR if possible and ideally, only make formatting decisions based on the shape of the AST. 

That's why I recommend using `default_specifier.is_none()` rather than `default.is_empty()`. 

I also find this easier to understand because I can reason about the formatting logic without having to know how `default` gets formatted.

File: crates/rome_js_formatter/src/formatter.rs [hunk 8]
```diff
@@ -107,6 +107,43 @@ impl Formatter {
         ])
     }
 
+    pub(crate) fn format_delimited_without_group(
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/formatter.rs:110]: Would you mind adding additional documentation of how this function is different from the others and in what situations it should be used?

**Review by MichaReiser — COMMENTED** [review]
Thanks for working on this. 

Have you checked [Prettier's](https://github.com/prettier/prettier/blob/d13feed42b6478710bebbcd3225ab6f203a914c1/src/language-js/print/module.js#L81-L91) implementation of the formatting. It seems to apply more rules than your implementation to determine if and when to break import statements. Can you explain how the two implementations differs and if your changes address all cases outlined in the issue. 

Please add some formatter tests that, at least, cover all the test cases of the linked issue.

**Comment by IWANABETHATGUY** [comment]
> 
Yeah, the prettier test case is already here https://github.com/rome/tools/pull/2498/files#diff-7903e95e4beeaeb0515e82814fac55ef523efb3d0336f775dd848db1a0a63becR16-R26, so I think I don't need to add any more, right?

File: crates/rome_js_formatter/tests/specs/prettier/js/import/inline.js.snap [hunk 9]
```diff
@@ -15,16 +16,19 @@ import {a2, somethingSuperLongsomethingSuperLong3} from 'somethingSuperLongsomet
 ```js
 import somethingSuperLongsomethingSuperLong from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
 import { somethingSuperLongsomethingSuperLong1 } from "somethingSuperLongsomethingSuperLongsomethingSuperLong";
```
Comment (leops) [inline crates/rome_js_formatter/tests/specs/prettier/js/import/inline.js.snap:18]: This test isn't verifying the case where the import specifier itself goes over the line length, we should add it to our own test suite:
```js
import {somethingSuperLongsomethingSuperLoooooooooooooooooooooooooooooooooooooooooooong} from 'source'
```

**Comment by ematipico** [comment]
Hi @IWANABETHATGUY , are there any updates on this PR? If you want and don't have time, we can take over the PR and try to finish it.

**Comment by IWANABETHATGUY** [comment]
> 

This was blocked by IR problem before, i would like to have a look today.

**Comment by IWANABETHATGUY** [comment]
@ematipico I am ready for review

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 10]
```diff
@@ -18,6 +21,52 @@ impl FormatNodeFields<JsImportNamedClause> for FormatNodeRule<JsImportNamedClaus
             assertion,
         } = node.as_fields();
 
+        let named_import = named_import?;
+
+        // can_break implementation, return `format_element` instead of boolean to reduce enum conversion overhead.
+        // if `can_break` is true we just use the previous format strategy, otherwise we use the new format strategy.
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:27]: I believe this comment is outdated. What's `can_break`? Probably it has to be revisited

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 11]
```diff
@@ -18,6 +21,52 @@ impl FormatNodeFields<JsImportNamedClause> for FormatNodeRule<JsImportNamedClaus
...
+            match named_import {
+                JsAnyNamedImport::JsNamedImportSpecifiers(ref specifiers)
+                    if specifiers.specifiers().len() == 1 =>
+                {
+                    let first_specifier = specifiers.specifiers().iter().next().unwrap();
+                    match first_specifier {
+                        Ok(JsAnyNamedImportSpecifier::JsShorthandNamedImportSpecifier(_)) => {
+                            let syntax_node = specifiers.clone().into_syntax();
+                            if syntax_node.has_leading_comments()
+                                || syntax_node.has_trailing_comments()
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:42]: You can use `has_comments_direct` instead

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 12]
```diff
@@ -18,6 +21,52 @@ impl FormatNodeFields<JsImportNamedClause> for FormatNodeRule<JsImportNamedClaus
...
+            named_import.format().format(formatter)
+        } else {
+            match named_import {
+                JsAnyNamedImport::JsNamedImportSpecifiers(ref specifiers)
+                    if specifiers.specifiers().len() == 1 =>
+                {
+                    let first_specifier = specifiers.specifiers().iter().next().unwrap();
+                    match first_specifier {
+                        Ok(JsAnyNamedImportSpecifier::JsShorthandNamedImportSpecifier(_)) => {
+                            let syntax_node = specifiers.clone().into_syntax();
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:40]: Do you need to consume `specifiers`? Isn't `specifiers.syntax()` enough?

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs (hunk 10 above)
Comment (ematipico) [reply crates/rome_js_formatter/src/js/module/import_named_clause.rs:27]: Could you please add some comment about the new format strategy? How does it work? 
Maybe some snippet inside the documentation would help.

It might be helpful to move this new logic inside a private function, so we you can add documentation comments,

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 13]
```diff
@@ -18,6 +21,52 @@ impl FormatNodeFields<JsImportNamedClause> for FormatNodeRule<JsImportNamedClaus
...
+        // reference https://github.com/prettier/prettier/blob/5b113e71b1808d6916f446c3aa49c3c53e3bdb98/src/language-js/print/module.js#L173
+        let formatted_named_import = if default_specifier.is_some() {
+            // `can_break` is true.
+            named_import.format().format(formatter)
+        } else {
+            match named_import {
+                JsAnyNamedImport::JsNamedImportSpecifiers(ref specifiers)
+                    if specifiers.specifiers().len() == 1 =>
+                {
+                    let first_specifier = specifiers.specifiers().iter().next().unwrap();
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:37]: Could you please add a `// SAFETY` comment, saying why it's safe to call `unwrap` here?

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 14]
```diff
@@ -18,6 +21,52 @@ impl FormatNodeFields<JsImportNamedClause> for FormatNodeRule<JsImportNamedClaus
             assertion,
         } = node.as_fields();
 
+        let named_import = named_import?;
+
+        // can_break implementation, return `format_element` instead of boolean to reduce enum conversion overhead.
+        // if `can_break` is true we just use the previous format strategy, otherwise we use the new format strategy.
+        // reference https://github.com/prettier/prettier/blob/5b113e71b1808d6916f446c3aa49c3c53e3bdb98/src/language-js/print/module.js#L173
+        let formatted_named_import = if default_specifier.is_some() {
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:29]: You could write this

```rs
if let Some(JsAnyNamedImport::JsNamedImportSpecifiers(ref specifiers)) = default_specifier {
    if specifiers.specifiers().len() == 1 {
        // .. rest of code
    } else {
       named_import.format().format(formatter) 
    }
} else {
    named_import.format().format(formatter)
}
```

Like this we remove a `match`

**Review by ematipico — CHANGES_REQUESTED** [review]
Thank you for looking into this! I'd request some changes, because we would like to document these particular formatting rules, especially if they come from prettier:

- document the new logic, especially with example and edge cases (comments?)
- add new tests that cover the new logic and new conditions

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs (hunk 14 above)
Comment (IWANABETHATGUY) [reply crates/rome_js_formatter/src/js/module/import_named_clause.rs:29]: It seems your code snippet is different from mine, `default_specifier` will never match `JsNamedImportSpecifiers` due to `default_specifier`'s type is `JsDefaultImportSpecifier`

**Comment by IWANABETHATGUY** [comment]
!bench_formatter

File: crates/rome_js_formatter/src/js/module/import_named_clause.rs [hunk 15]
```diff
@@ -18,6 +21,57 @@ impl FormatNodeFields<JsImportNamedClause> for FormatNodeRule<JsImportNamedClaus
...
+        // reference https://github.com/prettier/prettier/blob/5b113e71b1808d6916f446c3aa49c3c53e3bdb98/src/language-js/print/module.js#L173
+
+        // https://github.com/prettier/prettier/blob/5b113e71b1808d6916f446c3aa49c3c53e3bdb98/src/language-js/print/module.js#L184-L209v,
+        // `standaloneSpecifiers` corresponding our `default_specifier` + part of `named_import`,
+        // `groupedSpecifiers` corresponding our `named_import`
+        // https://github.com/prettier/prettier/blob/5b113e71b1808d6916f446c3aa49c3c53e3bdb98/src/language-js/print/module.js#L216-L243
+        // According L216-L219 we know that `canBreak` will be true only if 
+        let formatted_named_import = if default_specifier.is_some() {
+            // `can_break` is true.
+            named_import.format().format(formatter)
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/module/import_named_clause.rs:37]: Nit: You can use `formatted![formatter, [named_import.format()]]` reads a bit nicer.

**Comment by ematipico** [comment]
@IWANABETHATGUY how's the PR going? We would like to merge this PR for the next release, which is next week. Do you think you'll be able to address the CI failure and the feedbacks in the next few days?

If not, please let us know, so we can help.

File: crates/rome_js_formatter/tests/specs/prettier/js/import/comments.js.snap [hunk 16]
```diff
@@ -1,5 +1,6 @@
 ---
 source: crates/rome_js_formatter/tests/prettier_tests.rs
+assertion_line: 181
```
Comment (ematipico) [inline crates/rome_js_formatter/tests/specs/prettier/js/import/comments.js.snap:3]: This change can be reverted

**Review by ematipico — APPROVED** [review]
Looks good! Let's fix the conflict and then we can merge it
